### PR TITLE
Fix double locking sonar error

### PR DESCRIPTION
**Pull request by matthieun**
### Description:

[Sonar flags double locking with non-volatile members as a non thread-safe operation]( https://sonarcloud.io/project/issues?id=org.openstreetmap.atlas%3Aatlas&issues=AWSBg5Ghl6vpliCeK61L&open=AWSBg5Ghl6vpliCeK61L).
This PR attempts to solve this by using volatile spatial indices in `AbstractAtlas` and also refactors some methods to reduce code complexity.

Also avoided a code smell by using `M` instead of `Member` as a generic variable in many methods.

### Potential Impact:

Should have no impact.

### Unit Test Approach:

All unit tests covering de-serialization, and PackedAtlas and MultiAtlas creation pass.
DNM until large scale tests prove no other issue.

### Test Results:

DNM until large scale tests prove no other issue.

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/atlas/AbstractAtlas.java [hunk 1]
```diff
@@ -49,12 +57,17 @@
...
-    private transient SpatialIndex<Node> nodeSpatialIndex;
-    private transient SpatialIndex<Edge> edgeSpatialIndex;
-    private transient SpatialIndex<Area> areaSpatialIndex;
-    private transient SpatialIndex<Line> lineSpatialIndex;
-    private transient SpatialIndex<Point> pointSpatialIndex;
-    private transient SpatialIndex<Relation> relationSpatialIndex;
+    // Transient: Those are not serialized, and re-generated on the fly
+    // Volatile: This is to allow double checked locking to be safe in the
+    // this.buildXXXXXSpatialIndexIfNecessary() methods.
+    // http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/AbstractAtlas.java]: Good read, thanks for linking.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/Atlas.java [hunk 2]
```diff
@@ -185,6 +185,19 @@
...
+     *
+     * @param type
+     *            The type to restrain to
+     * @param memberClass
+     *            The class of the member
+     * @param <M>
+     *            The AtlasEntity type
+     * @return All the {@link AtlasEntity}s of a specific type
+     */
+    <M extends AtlasEntity> Iterable<M> entities(ItemType type, Class<M> memberClass);
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/Atlas.java]: Nit: One quick test in `BareAtlasTest` for this?

Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/Atlas.java]: Added!

**Comment by matthieun** [comment]
Verified that it still works at scale. This should be good to merge.

**Pull request merged**